storage/src/merkle_tree_db_adapter.cpp

```diff
@@ -0,0 +1,323 @@
+  for (const auto &intKey : batch.stale.internal_keys) {
+    const auto key = DBKeyManipulator::genStaleDbKey(intKey, batch.stale.stale_since_version.value());
+    updates[key] = emptySliver;
+  }
+
+  // Leaf stale node indexes. Use the index only and set the key to an empty sliver.
+  for (const auto &leafKey : batch.stale.leaf_keys) {
+    const auto key = DBKeyManipulator::genStaleDbKey(leafKey, batch.stale.stale_since_version.value());
+    updates[key] = emptySliver;
+  }
+
+  // Internal nodes.
+  for (const auto &[intKey, intNode] : batch.internal_nodes) {
+    const auto key = DBKeyManipulator::genInternalDbKey(intKey);
+    updates[key] = serializeImp(intNode);
+  }
+
+  // Leaf nodes.
+  for (const auto &[leafKey, leafNode] : batch.leaf_nodes) {
+    const auto key = DBKeyManipulator::genDataDbKey(leafKey);
+    updates[key] = serializeImp(leafNode);
+  }
+
+  return updates;
+}
+
+}  // namespace
+
+Sliver DBKeyManipulator::genBlockDbKey(BlockId version) { return serialize(EDBKeyType::Block, version); }
+
+Sliver DBKeyManipulator::genDataDbKey(const LeafKey &key) { return serialize(EKeyType::Leaf, key); }
+
+Sliver DBKeyManipulator::genDataDbKey(const Sliver &key, BlockId version) {
+  auto hasher = Hasher{};
+  return serialize(EKeyType::Leaf, LeafKey{hasher.hash(key.data(), key.length()), version});
+}
+
+Sliver DBKeyManipulator::genInternalDbKey(const InternalNodeKey &key) { return serialize(EKeyType::Internal, key); }
+
+Sliver DBKeyManipulator::genStaleDbKey(const InternalNodeKey &key, BlockId staleSinceVersion) {
+  // Use a serialization type to discriminate between internal and leaf keys.
+  return serialize(EKeyType::Stale, staleSinceVersion, KeySerializationType::Internal, key);
+}
+
+Sliver DBKeyManipulator::genStaleDbKey(const LeafKey &key, BlockId staleSinceVersion) {
+  // Use a serialization type to discriminate between internal and leaf keys.
+  return serialize(EKeyType::Stale, staleSinceVersion, KeySerializationType::Leaf, key);
+}
+
+Sliver DBKeyManipulator::generateMetadataKey(ObjectId objectId) { return serialize(EBFTType::Metadata, objectId); }
+
+Sliver DBKeyManipulator::generateStateTransferKey(ObjectId objectId) { return serialize(EBFTType::ST, objectId); }
+
+Sliver DBKeyManipulator::generateSTPendingPageKey(uint32_t pageId) {
+  return serialize(EBFTType::STPendingPage, pageId);
+}
+
+Sliver DBKeyManipulator::generateSTCheckpointDescriptorKey(uint64_t chkpt) {
+  return serialize(EBFTType::STCheckpointDescriptor, chkpt);
+}
+
+Sliver DBKeyManipulator::generateSTReservedPageStaticKey(uint32_t pageId, uint64_t chkpt) {
+  return serialize(EBFTType::STReservedPageStatic, pageId, chkpt);
+}
+
+Sliver DBKeyManipulator::generateSTReservedPageDynamicKey(uint32_t pageId, uint64_t chkpt) {
+  return serialize(EBFTType::STReserverdPageDynamic, pageId, chkpt);
+}
+
+BlockId DBKeyManipulator::extractBlockIdFromKey(const concordUtils::Key &key) {
+  const auto offset = key.length() - sizeof(BlockId);
```

Comment:
This works, but any change to the LeafKey structure or addition of another key type will break it. Maybe add a comment that this only works for a serialized `LeafKey`? Later on we should have round trip tests so we can catch breakages in deserialization.

Reply:
Yeah, that's a good point. I only kept it to match up with the existing interface. As of now, it should work for block and leaf keys. I think we should not have that function in the future (or at least not in that form). I will add the comment for now.